models/models.go

```diff
@@ -139,6 +139,26 @@ func init() {
 	}
 }
 
+// WithEngine represents executing database operations
+func WithEngine(f func(e Engine) error) error {
```

Comment:
OK, although this is kind of what I was thinking we should do - it's not quite what I was thinking.

By passing in a `models.Engine` you're allowing arbitrary use of the xorm API outside of models - which I don't think we should permit. Use of xorm and the db probably should be restricted to functions in models.

I think we should be passing in a context (or at least a context provider). Something like:

```go
type DBContext {
  e Engine
}
```

then we could have `WithoutTransaction` and `WithTransaction` callback functions that set the `DBContext` appropriately. Then everywhere that uses `models.x` would be called within at least a `WithoutTransaction`. 

A better alternative would be through the use of a DBContextProvider interface:
```
type DBContextProvider interface {
    GetDBContext() DBContext
}
```

We make our Gitea and API contexts a `DBContextProvider` by setting their `DBContext` to be say the `models.DefaultDBContext` - which simply returns `models.x` as the little `e` Engine. We then make any call to an exported function in `models` expect a `DBContextProvider` as the first argument.

`models.WithTransaction()` then provides a callback with a different `DBContextProvider` which has a transaction in it.